include/pybind11/common.h

```diff
@@ -655,6 +655,11 @@ static constexpr detail::overload_cast_impl<Args...> overload_cast = {};
 ///  - sweet:   overload_cast<Arg>(&Class::func, const_)
 static constexpr auto const_ = std::true_type{};
 
+#else // no overload_cast: providing something that static_assert-fails:
+template <typename... Args> struct overload_cast {
+    static_assert(std::is_same<int, detail::void_t<Args...>>::value,
+                  "pybind11::overload_cast<...> requires compiling in C++14 mode");
```

Comment:
Looks good to me. Minor nitpick: `deferred_t` is tailor-made for situations like this: `deferred_t<std::false_type, Args...>::value`.